DREAM/Optimization/tsgGradientDescent.cpp

```diff
@@ -0,0 +1,126 @@
+
+    OptimizationStatus status;
+    status.num_iterations = 0;
+    status.stationarity_residual = std::numeric_limits<double>::max();
+
+    int num_dimensions = state.num_dimensions;
+    std::vector<double> &x = state.getXRef();
+    std::vector<double> x_prev(x), g_at_x_prev(g(x_prev)), g_at_x(num_dimensions), next_x(num_dimensions);
+    double f_at_x_prev(f(x_prev)), f_at_x(0);
+
+    while(status.num_iterations < num_iterations) {
+        // Find the next stepsize/candidate point by making sure it satisfies the well-known descent inequality (see γ_u in the
+        // reference paper associated with this function).
+        double lhs = 0;
+        double rhs = 0;
+        do {
+            if (status.num_iterations >= num_iterations)
+                break;
+            for (int j=0; j<num_dimensions; j++)
+                x[j] = x_prev[j] - g_at_x_prev[j] * state.adaptive_stepsize;
+            x = proj(x);
+            f_at_x = f(x);
+            lhs += f_at_x - f_at_x_prev;
+            for (int j=0; j<num_dimensions; j++) {
+                double delta = x[j] - x_prev[j];
+                rhs += delta * delta / (2.0 * state.adaptive_stepsize);
+                lhs -= g_at_x_prev[j] * delta;
+            }
+            state.adaptive_stepsize /= line_search_coeffs[0];
+            status.num_iterations++;
+        } while (lhs > rhs + TasGrid::Maths::num_tol);
+        state.adaptive_stepsize *= line_search_coeffs[0]; // offset the do-while loop.
+        // Check for approximate stationarity.
+        g_at_x = g(x);
+        status.stationarity_residual = compute_stationarity_residual(x, x_prev, g_at_x, g_at_x_prev, state.adaptive_stepsize);
+        if (status.stationarity_residual <= tolerance)
+            break;
+        // Update the key iterates manually for the next loop.
+        std::swap(x_prev, x);
+        std::swap(f_at_x_prev, f_at_x);
+        std::swap(g_at_x_prev, g_at_x);
+        // Optimistic stepsize update (see γ_d in the referenced paper above).
+        state.adaptive_stepsize *= line_search_coeffs[1];
+    }
+    return status;
+    
+}
+
+OptimizationStatus GradientDescent(const ObjectiveFunctionSingle &f, const GradientFunctionSingle &g, GradientDescentState &state,
+                                   const std::vector<double> &line_search_coeffs,  const int num_iterations, const double tolerance) {
+    // Wrapper to the proximal version with projection function == identity function.
+    return GradientDescent(f, g, identity, state, line_search_coeffs, num_iterations, tolerance);
+};
+
+OptimizationStatus GradientDescent(const GradientFunctionSingle &g, std::vector<double> &x, const double stepsize, const int num_iterations,
+                                   const double tolerance) {
+    OptimizationStatus status;
+    status.num_iterations = 0;
+    status.stationarity_residual = std::numeric_limits<double>::max();
+    std::vector<double> x_prev(x), g_at_x_prev(g(x_prev)), g_at_x(x.size());
```

Comment:
Naming convention? I don't like the `_at_` portion of the name, `g_at_x` has 5 characters, but only `g` and `x` have a meaning, the mid part is just clutter to differentiate from the name of the input variable.

I know you need to differentiate between `g_at_x` and the GradientFunctionSingle `g`, but how about calling things:
```
OptimizationStatus GradientDescent(const GradientFunctionSingle &grad, std::vector<double> &state, ...) {}
```
Then you can have variables like `x`, `gx` and `gstate`, where `x` is the internal temp and `state` is the result. The variables are easy to differentiate and the inputs are more expressive.

Similarly `f` can be `objective` or `cost` or `func` (for functional)

Reply:
Changed. `f` is now `func`, `g` is now `grad`, and the `_at_` portion has been removed in all variables.